// @tanstack/react-form/dist/esm/start/createServerValidate.js
import { decode } from "decode-formdata";
import { _tanstackInternalsCookie } from "./utils.js";
import { ServerValidateError } from "./error.js";
const createServerValidate = (defaultOpts) => async (ctx, formData, info) => {
  const { validatorAdapter, onServerValidate } = defaultOpts;
  const runValidator = async (propsValue) => {
    if (validatorAdapter && typeof onServerValidate !== "function") {
      return validatorAdapter().validateAsync(propsValue, onServerValidate);
    }
    return onServerValidate(propsValue);
  };
  const referer = ctx.request.headers.get("referer");
  const data = decode(formData, info);
  const onServerError = await runValidator({ value: data });
  if (!onServerError) return;
  const formState = {
    errorMap: {
      onServer: onServerError
    },
    values: data,
    errors: onServerError ? [onServerError] : []
  };
  const cookie = await _tanstackInternalsCookie.serialize(formState);
  throw new ServerValidateError({
    response: new Response("ok", {
      headers: {
        Location: referer,
        "Set-Cookie": cookie
      },
      status: 302
    }),
    formState
  });
};
export {
  createServerValidate
};
