// @tanstack/react-form/dist/cjs/start/createServerValidate.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const decodeFormdata = require("decode-formdata");
const utils = require("./utils.cjs");
const error = require("./error.cjs");
const createServerValidate = (defaultOpts) => async (ctx, formData, info) => {
  const { validatorAdapter, onServerValidate } = defaultOpts;
  const runValidator = async (propsValue) => {
    if (validatorAdapter && typeof onServerValidate !== "function") {
      return validatorAdapter().validateAsync(propsValue, onServerValidate);
    }
    return onServerValidate(propsValue);
  };
  const referer = ctx.request.headers.get("referer");
  const data = decodeFormdata.decode(formData, info);
  const onServerError = await runValidator({ value: data });
  if (!onServerError) return;
  const formState = {
    errorMap: {
      onServer: onServerError
    },
    values: data,
    errors: onServerError ? [onServerError] : []
  };
  const cookie = await utils._tanstackInternalsCookie.serialize(formState);
  throw new error.ServerValidateError({
    response: new Response("ok", {
      headers: {
        Location: referer,
        "Set-Cookie": cookie
      },
      status: 302
    }),
    formState
  });
};
exports.createServerValidate = createServerValidate;
