// @tanstack/react-form/dist/cjs/nextjs/createServerValidate.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const decodeFormdata = require("decode-formdata");
const error = require("./error.cjs");
const createServerValidate = (defaultOpts) => async (formData, info) => {
  const { validatorAdapter, onServerValidate } = defaultOpts;
  const runValidator = async (propsValue) => {
    if (validatorAdapter && typeof onServerValidate !== "function") {
      return validatorAdapter().validateAsync(propsValue, onServerValidate);
    }
    return onServerValidate(propsValue);
  };
  const values = decodeFormdata.decode(formData, info);
  const onServerError = await runValidator({ value: values });
  if (!onServerError) return;
  const formState = {
    errorMap: {
      onServer: onServerError
    },
    values,
    errors: onServerError ? [onServerError] : []
  };
  throw new error.ServerValidateError({
    formState
  });
};
const initialFormState = {
  errorMap: {
    onServer: void 0
  },
  values: void 0,
  errors: []
};
exports.createServerValidate = createServerValidate;
exports.initialFormState = initialFormState;
